sql, backupccl: backup and restore fk-upgrade-downgrade awareness

jordanlewis opened the pull request:
This commit has a solution for all 4 cases that you need to worry about
with regards to writing and reading table descriptors in a mixed
19.1/19.2 state.

We now:

1. upgrade the descriptors from disk while creating a backup
2. downgrade the descriptors from memory into the backup descriptor if
   we're in a 19.1/2 mixed state
1. upgrade the descriptors from a backup while preparing to restore
4. downgrade the descriptors from memory onto the disk while restoring
   a backup if we're in a 19.1/2 mixed state

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/39474)
<!-- Reviewable:end -->

lucy-zhang committed:
sql, backupccl: backup and restore fk-upgrade-downgrade awareness

This commit has a solution for all 4 cases that you need to worry about
with regards to writing and reading table descriptors in a mixed
19.1/19.2 state.

We now:

1. upgrade the descriptors from disk while creating a backup
2. downgrade the descriptors from memory into the backup descriptor if
   we're in a 19.1/2 mixed state
1. upgrade the descriptors from a backup while preparing to restore
4. downgrade the descriptors from memory onto the disk while restoring
   a backup if we're in a 19.1/2 mixed state

Release note: None

Co-authored-by: Lucy Zhang <lucy@cockroachlabs.com>
Co-authored-by: Jordan Lewis <jordan@cockroachlabs.com>
